### PR TITLE
Retain comments inside return statements

### LINKED ISSUE #17532: Typescript eats Google Closure type cast comment
**TypeScript Version:**  2.4.20

**Code**

```ts
function f() {
    return /* @type {number} */ 42;
}
```

**Expected behavior:**
Should be compiled to
```js
function f() {
    return /* @type {number} */ 42;
}
```

**Actual behavior:**
Compiled to
```js
function f() {
    return 42;
}
```

### PR DESCRIPTION
**Pull request by weswigham**
by including the return keyword in the parse tree, so its position is recorded for comment emit.

A simple
```ts
            if (node.expression) {
                emitLeadingCommentsOfPosition(node.expression.pos);
            }
```
inside `emitReturnStatement` is unsuitable, since multiline-inline comments are not collected as potential leading comments of a position (instead, they get collected as trailing comments for a prior token - in this case, the `ReturnKeyword` we never recorded the position of), and there's no point of reference to call `emitTrailingCommentsOfPosition` with.

Fixes #17532

### CODE REVIEW

**Comment by DanielRosenwasser** [comment]
I am slightly concerned about adding another token node for every `return` statement. At what point does the tree become concrete for all terminal nodes? Is that a cost we're willing to pay?

**Comment by weswigham** [comment]
@DanielRosenwasser I think I could get the correct position by getting the statement start, skipping trivia, then adding "return".length; but that's another special case in the emitter for return statements, which also should be evaluated. If it works (it should?), then I'm OK swapping to doing it that way; its just a fair bit of late calculation for something we already recorded way back in the parser, but decided not to keep.

It's a matter of AST size vs emit time (a small delta, either way); so which is preferable to minimize?

**Comment by rbuckton** [comment]
> I think I could get the correct position by getting the statement start, skipping trivia, then adding "return".length

This is the approach we took for PropertyAccess and it works well enough. Currently, `emitReturnStatement` uses `writeToken` which already does position tracking for source maps. We could augment that to emit comments as well, and possibly end up with higher fidelity for comment emit in other places.

File: src/compiler/types.ts [hunk 1]
```diff
@@ -1750,6 +1750,7 @@ namespace ts {
 
     export interface ReturnStatement extends Statement {
         kind: SyntaxKind.ReturnStatement;
+        returnKeyword: Token<SyntaxKind.ReturnKeyword>;
```
Comment (rbuckton) [inline src/compiler/types.ts]: I'd rather not increase the size of the node. Instead, consider modifying `writeTokenText` in emitter to emit comments as well as source maps. 

Keep in mind that `writeTokenText` is used quite a bit throughout the emitter. Changes could have an impact on emit performance. If the impact is minimal that change would be preferable. If it is significant then we either need to find ways to mitigate the cost or consider the approach in the current PR.

**Comment by scf37** [comment]
This issue seems to affect all rhs expressions:

```typescript
function f() {
  var a = /* foo */ 1;
  var b = (1 + /* bar */ a);
  return /* foobar */ (a + b);
}
```

compiles to 

```javascript
function f() {
    var a = 1;
    var b = (1 + a);
    return (a + b);
}
```

**Comment by weswigham** [comment]
@scf37 It's not all RHS expressions, but rather all trailing comments of a token which appears in the middle of a parsed node. Once this fix is merged, I plan on looking into doing a similar operation to try to retain comments on other such tokens.

**Comment by rbuckton** [comment]
@weswigham Is there an issue for the other token positions where we don't preserve comments? If not, can you add one?

**Comment by weswigham** [comment]
Opened - I was able to find a general issue about generally better comment emit - #2546 - but nothing specific to this area.

**Pull request merged**